app/services/facebook.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+class Facebook
+  include HTTParty
+
+  class << self
+    def find_user_by_access_token(access_token)
+      user_id = validate_and_extract_user_id(access_token)
+      info = fetch_account_profile(access_token, user_id)
+      find_and_update_user(info) || create_user(info)
+    rescue JSON::ParserError => e
+      raise Facebook::FailedAuthentication, e.message
+    end
+
+    private
+
+    def find_and_update_user(user_info)
+      user = User.find_by(facebook_user_id: user_info[:facebook_user_id])
+      return nil unless user
+      user.update(user_info)
+      user
+    end
+
+    def create_user(user_info)
+      User.create!(user_info)
+    end
+
+    def validate_and_extract_user_id(access_token)
+      unless /^[0-9a-zA-Z]+$/.match?(access_token)
```

Comment:
just curious, did they specify somewhere in their documentation that facebook access_tokens follow this format?

Reply:
I couldn't find anything on that, just from observation of your token. I guess that's a pretty small sample heh. I'm just concerned about the security hole of someone adding & or ? in their token and then they're able to pass other params.

Reply:
rails doesn't provide any sort of uri escaping functionality to allow you to set url parameters safely no matter what characters they contain?

Reply:
it looks like in order to get escaping you should be sending the query parameters as a second parameter.

https://github.com/jnunemaker/httparty#examples

Reply:
@andrewroth bumping my response :)